### ISSUE TITLE
Some questions.

### ISSUE THREAD

**Issue by kfsdi** [AlexanderBabansky/EasyRTMP#1]
Hello! I liked your library, but I have a question:
How can I, being a client, connect to an RTMP server and start sending packets? Can you provide me with the code? I've been trying to come up with something for several hours, but I don't fully understand how your library works.

**Comment by AlexanderBabansky** [comment]
there is an [example](https://github.com/AlexanderBabansky/RTMP-Restreamer/blob/main/restreamer.cpp) program that restreams media. It starts server, initialize client and connect to thirt party server, and just pass-through opaque data.
Note, if you want to stream your own encoded media, it must be encoded with supported codec for receiving side (for video usually h264, for audio aac). Also video must be in a valid bitstream format, for h264 NALU must be **not in AnnexB**, but in **AVCC** format, just like in FLV.
[Here](https://github.com/AlexanderBabansky/WebdriverTorso/blob/master/main.cpp) is some other example, but unfortunately it is not ready for building.

**Comment by kfsdi** [comment]
No, I'm not transmitting audio or video, I'm just trying to send commands to the server. I saw your example, but I don't understand, is a separate server being created here to receive messages from the thirdparty server and a client is being created that sends data there? Are these not 2 different connections? I just need to receive and send packets from the same socket

**Comment by kfsdi** [comment]
If it's not difficult for you, write the code how the program that connects me to the server and produces a handshake should look like, and gives me the opportunity to send packets to the server 
I will be very grateful

**Comment by kfsdi** [comment]
This server has an rtmp address: rtmp://37.112.251.100:3082 (This is not my server, but the server of the game)

**Comment by AlexanderBabansky** [comment]
>I'm just trying to send commands to the server

What kind of commands do you want to send to game server?

**Comment by kfsdi** [comment]
It's 0x14 (AMF0) packets.
Show me an example of how to connect to the server, make a handshake and send AMF0 Packet

**Comment by AlexanderBabansky** [comment]
Here is code for sending simple AMF command. The handshake is implemented in **RTMPEndpoint** class.
```C
TCPClient tcp_client;
std::shared_ptr<TCPNetwork> tcp_network;
try {
    tcp_network = tcp_client.ConnectToHost(ipAddress, port);
    librtmp::RTMPEndpoint rtmp_endpoint(tcp_network.get());
    AMFValue amf_container(AMFType::ROOT);
    AMFValue amf_cmd_name(AMFType::STRING);
    amf_cmd_name.set_string("connect");
    AMFValue amf_transaction_id(AMFType::NUMBER);
    amf_transaction_id.set_number(1);
    amf_container.push_back(amf_cmd_name);
    amf_container.push_back(amf_transaction_id);
    rtmp_endpoint->SendCommand(amf_container);
}
catch (std::exception& e) {
    ...
}
```
Also note, that you have to receive server responses, and pass service messages (not AMF) to endpoint, and handle your AMF responses. Example is in [RTMPClientSession::GetCommandResponse](https://github.com/AlexanderBabansky/EasyRTMP/blob/main/rtmp_client_session.cpp#L8)

**Comment by kfsdi** [comment]
Thank you very much! It works!!

**Comment by kfsdi** [comment]
Please tell me how to fill in ECMA_ARRAY? I tried different parameters, but got exception in AMFValue::push_back

**Comment by kfsdi** [comment]
Hello, how do I read the received RTMP Packets? In one of the files I found:
```as 
        while (true)
        {
            if (GetCommandResponse(
                [&](AMFValue v) {
                    auto it = v.begin();
                    if ((*it).get_string() == "_result") {
                        ++it;
                        ++it;
                        ++it;
                        if ((*it)["code"].get_string() == "NetConnection.Connect.Success") {
                            return true;
                        }
                    }
                    return false;
                },
                1, rtmp_endpoint)
                == false) {
                throw rtmp_exception(RTMPError::NETSTREAM_CONNECTION);
            }
            break;
        }
```
And it works. But when I want to apply this to another package, it doesn't work. Example of packet:
![image](https://github.com/AlexanderBabansky/EasyRTMP/assets/152034514/a265a3b4-1fb7-41a4-96f0-d6c8743064d8)

**Comment by kfsdi** [comment]
i got
Assertion failed: false, file D:\projekt\ConsoleApplication2\amf.cpp, line 181

**Comment by kfsdi** [comment]
Now i tried:
`        RTMPMessageType message_type = RTMPMessageType::UNKNOWN;
        uint32_t message_stream_id = 0;
        uint32_t message_length = 0;
        uint64_t timestamp = 0;
        uint8_t chunk_stream_id = 0;
        auto data = rtmp_endpoint->GetRTMPMessage(&message_type, &message_stream_id, &message_length,
            &chunk_stream_id, &timestamp);
        for (int i = 0; i < data.size(); ++i)
        {
            std::cout << data[i];
        }
        return 1;
`

And i got:
`
_DUI@Title╨Э╨╡╨┤╨╡╨╗╤П ╨┐╨╛╨║╤Г╨┐╨╛╤З╨╡╨║Reward5IconMRId@хо WeekId│hScore
LastWeekDate
Bx└с6Reward3HintTRId@с╡IconMRId@╩тАReward4HintTRIdTasks
StackId?ЁEventId@2
EventCount@(
EventObjectIIconMRId@╩╣АIsDoneRId
EventValuTextTRId@уz TaskId@°pp
LocationIHintTRId@у}а   StackId@EventId@;
EventCount@
EventObjectId@IconMRId@╦IsDoneRId
EventValuTextTRId@уx@TaskIdm0
LocationId@HintTRId@у~@ StackIdEventId@A
EventCount@.
EventObjectId└YIconMRId@╩╗АIsDone
EventValuTextTRId@у{аTaskIdARд
LocationId@HintTRId@у~А StackId@EventId@7
EventCount?Ё
EventObjectIIconMRId@╩╝АIsDoned
EventValuTextTRId@у|└TaskIdAkФ
LocationId@UHintTRId@у~└        Reward4IconMRIDaysLeft@0*6К7$Reward5HintTRId@у Reward3IconMRId@хнр
`
how can i parse this?

**Comment by AlexanderBabansky** [comment]
I said about receiving messages [code](https://github.com/AlexanderBabansky/EasyRTMP/blob/main/rtmp_client_session.cpp#L8), this method expects strict response semantic with NumberID (second AMF value).  assertion failing means corrupted, malformed, or not default AMF response. Parsing was tested with public servers, like Facebook, Twitch, Youtube, Instagram, TikTok, nginx and it works fine with them. You should debug and find which AMF value it can't parse.
We have the same problems with Facebook, it produces not compliant AMF arrays, but I fixed with workaround.

**Comment by kfsdi** [comment]
Sorry for my stupid questions, but I don't even understand the specification, but how to send a Ping Response (User Control Message Event 7).
